controllers/catalogsource_controller.go

```diff
@@ -0,0 +1,158 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/deppy/pkg/deppy"
+	"github.com/operator-framework/deppy/pkg/deppy/input"
+	"github.com/operator-framework/operator-controller/internal/resolution/entity_sources/catalogsource"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	defaultCatalogSourceSyncInterval     = 5 * time.Minute
+	defaultRegistryGRPCConnectionTimeout = 10 * time.Second
+
+	eventTypeNormal  = "Normal"
+	eventTypeWarning = "Warning"
+
+	eventReasonCacheUpdated      = "BundleCacheUpdated"
+	eventReasonCacheUpdateFailed = "BundleCacheUpdateFailed"
+)
+
+type CatalogSourceReconciler struct {
+	sync.RWMutex
+	client.Client
+	scheme   *runtime.Scheme
+	registry catalogsource.RegistryClient
+	recorder record.EventRecorder
+	cache    map[string]map[deppy.Identifier]*input.Entity
+}
+
+func NewCatalogSourceReconciler(client client.Client, recorder record.EventRecorder) *CatalogSourceReconciler {
+	return &CatalogSourceReconciler{
+		RWMutex:  sync.RWMutex{},
+		Client:   client,
+		recorder: recorder,
+		cache:    map[string]map[deppy.Identifier]*input.Entity{},
+		registry: catalogsource.NewRegistryGRPCClient(defaultRegistryGRPCConnectionTimeout),
+	}
+}
+
+// +kubebuilder:rbac:groups=operators.coreos.com,resources=catalogsources,verbs=get;list;watch
+
+func (r *CatalogSourceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx).WithName("catalogsource-controller")
+	l.V(1).Info("starting")
+	defer l.V(1).Info("ending")
+
+	var catalogSource = &v1alpha1.CatalogSource{}
+	if err := r.Client.Get(ctx, req.NamespacedName, catalogSource); err != nil {
+		if errors.IsNotFound(err) {
+			r.dropSource(req.String())
```

Comment:
We are deleting the catsrc immediately if we are not able to fetch it from cluster. I'm wondering given this use case, would it make sense to use the [uncached client ](https://github.com/kubernetes-sigs/controller-runtime/blob/ffb74e54a54bcb10a615c501416820e06051b9b4/pkg/client/client.go#L76)for this purpose so that we don't run into additional reconciles for stale c-r caches.

@joelanford would that make more sense, instead of using a cached client and redoing the work of converting to entities if there is a stale cache. I doubt we will hit any performance issues without caching for this specific case of watching catsrc?

Reply:
Since the controller is currently watching _only_ catalog sources, I don't think the cache will never be stale. I _think_ the cached client is fine.

You bring up a good point about redoing work of listing entities though. It's a bummer that there's no GRPC endpoint that can tell us if we _need_ to relist. This is something we should probably handle in the new OLMv1 catalog source spec (cc @anik120).

Reply:
Another naive question: As an OLM user would I in future be allowed to specifically select the catsrc I want to restrict for resolution? Probably through annotations? 

For now we are watching all the catalog sources in the namespace. Is it necessary, if the user wants to use only a specific set of catsrc for resolution. And how do we provide the user that option. If its through Operator object, then we would need to dynamically change the watches after catsrc controller has started.

Or does we make this controller to watch all the catsrc objects, and do that filtering in the operator controller when it takes in the entities for resolution?

Reply:
@anik120 Follow up question on that. If not the user, does the admin have the privilege of not selecting a particular catalogsrc to install dependencies from - for an operator? 

The reason for asking that was from the perspective of whether we need to watch all catalogsources in the cluster necessarily.

But now my other question is - is there a particular design reason why we would not allow someone (whether its admin or any other persona) to not select the catsrc for resolution (Say blacklist a catsrc from being used for resolution)? 

Also, since you mentioned that we would not want to go in that path (this wouldn't affect the watches part of it, but just asking to understand if there is a particular reason):

> "when you as a tenant want to install an operator, you can specify which CatalogSources you want to install the dependencies from".

why not? Say I don't trust the publisher of this catsrc, I don't want to use its dependencies for resolution.

Reply:
@varshaprasad96 @anik120 I'm not sure the admin can disable a catalogsource so it can be ignored when installing dependencies, apart from deleting the catalogsource. Since the admin is the one that creates and manages what catalogsources are available on cluster, it makes sense to me that they can also decide their priority when resolving something to be installed on cluster. 

CatalogSource priorities are currently declared for each catalogSource, so this priority doesn't change each time we perform resolution. If we had tenants declaring their priorities, it could get potentially very messy, especially since two tenants could declare different priorities for operators being resolved at the same time.